pkg/controller/job/job_controller_test.go

```diff
@@ -5253,7 +5253,7 @@ func TestFinalizerCleanup(t *testing.T) {
 	defer cancel()
 
 	clientset := fake.NewSimpleClientset()
-	sharedInformers := informers.NewSharedInformerFactory(clientset, controller.NoResyncPeriodFunc())
+	sharedInformers := informers.NewSharedInformerFactory(clientset, controller.StaticResyncPeriodFunc(1*time.Second)())
```

Comment:
This is not what the production code does, so it might just hide other bugs.

However, in production we don't disable workers for job sync or orphan pods, which I think would have caught the "missing" Pod.

A more robust hint is to call `manager.expectations.SatisfiedExpectations` in the test, after creating the pod.

Reply:
Could you please link me to an example, as I am not fully familiar with this logic?

Reply:
https://github.com/kubernetes/kubernetes/blob/36930f4cb605450ccbe3126caad365aa5f51267c/pkg/controller/job/job_controller.go#L773

This map simply tells you whether the event handler has seen all created pods.... Although, you might need to increment the counter before creating the Pod (this is normally done by syncJob).

https://github.com/kubernetes/kubernetes/blob/36930f4cb605450ccbe3126caad365aa5f51267c/pkg/controller/job/job_controller.go#L1535

Reply:
I've done pattern matching and moved locally 
```
	// Start the Pod and Job informers.
	sharedInformers.Start(ctx.Done())
	sharedInformers.WaitForCacheSync(ctx.Done())
```
before `go manager.Run`, as in other unit tests. This seems to fix the issue, but I don't exactly know why.

Reply:
I tried doing something like:
```
	job := newJob(1, 1, 1, batch.NonIndexedCompletion)
	job, err = clientset.BatchV1().Jobs(job.GetNamespace()).Create(ctx, job, metav1.CreateOptions{})
	if err != nil {
		t.Fatalf("Creating job: %v", err)
	}

	key, err := controller.KeyFunc(job)
	if err != nil {
		t.Errorf("Unexpected error getting job key: %v", err)
	}
	if err := manager.expectations.ExpectCreations(logger, key, 1); err != nil {
		t.Errorf("Unexpected error setting creation expectations: %v", err)
	}

	// Create a Pod with the job tracking finalizer
	pod := newPod("test-pod", job)
	pod.Finalizers = append(pod.Finalizers, batch.JobTrackingFinalizer)
	pod, err = clientset.CoreV1().Pods(pod.GetNamespace()).Create(ctx, pod, metav1.CreateOptions{})
	if err != nil {
		t.Fatalf("Creating pod: %v", err)
	}

	if err := wait.PollUntilContextTimeout(ctx, 100*time.Millisecond, wait.ForeverTestTimeout, true, func(ctx context.Context) (bool, error) {
		return manager.expectations.SatisfiedExpectations(logger, key), nil
	}); err != nil {
		t.Errorf("Waiting for Job expecations to get satisfied: %v", err)
	}
```

But now after ~35s I get the following error:
```
I1114 15:51:42.267620   62752 shared_informer.go:311] Waiting for caches to sync for job
I1114 15:51:42.267638   62752 shared_informer.go:318] Caches are synced for job
--- FAIL: TestFinalizerCleanup (30.20s)
    job_controller.go:224: I1114 15:51:42.267370] Starting job controller
    controller_utils.go:224: I1114 15:51:42.367794] Setting expectations expectations={}
    controller_utils.go:196: I1114 15:51:42.367948] Controller still waiting on expectations expectations={}
    job_controller.go:554: I1114 15:51:42.368107] enqueueing job key="default/foobar"
    controller_utils.go:196: I1114 15:51:42.468142] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:42.573333] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:42.669239] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:42.771465] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:42.869245] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:42.968475] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:43.069376] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:43.169245] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:43.268136] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:43.368991] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:43.468583] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:43.569034] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:43.668552] Controller still waiting on expectations expectations={}
    controller_utils.go:196: I1114 15:51:43.768183] Controller s
```

Reply:
And if I try just setting the `ExpectedCreations` to 1 in the `TestFinalizerCleanup` case without a call to `SatisfiedExpectations`, I get this error:
```
I1114 16:26:50.771182   70807 shared_informer.go:311] Waiting for caches to sync for job
I1114 16:26:50.771248   70807 shared_informer.go:318] Caches are synced for job
--- FAIL: TestFinalizerCleanup (30.10s)
    job_controller.go:224: I1114 16:26:50.771018] Starting job controller
    controller_utils.go:224: I1114 16:26:50.871967] Setting expectations expectations={}
    job_controller.go:554: I1114 16:26:50.872239] enqueueing job key="default/foobar"
    job_controller.go:554: I1114 16:26:50.872292] enqueueing job key="default/foobar"
    controller_utils.go:241: I1114 16:26:50.872447] Lowered expectations expectations={}
    job_controller.go:554: I1114 16:26:50.872484] enqueueing job key="default/foobar"
    job_controller_test.go:5320: Waiting for Pod to get the finalizer removed: context deadline exceeded
FAIL
```

It correctly sees the expectation, creates the Pod, lowers the expectation, but the Pod informer does not have it in its cache.

Who is responsible for updating the Pod informer cache? I fail to understand what happens under load so it doesn't have it in the cache?

Reply:
> This is not what the production code does, so it might just hide other bugs.
> 
> However, in production we don't disable workers for job sync or orphan pods, which I think would have caught the "missing" Pod.
> 
> A more robust hint is to call `manager.expectations.SatisfiedExpectations` in the test, after creating the pod.

I see that when Kubernetes starts controllers, they have a `ResyncPeriod` - https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L584

What could happen if we use a `ResyncPeriod` in the test? @alculquicondor

Reply:
Re @mimowo
We should wait for caches to sync before starting the manager regardless. It's a good practice to only start the workers once we know that the cache is fully in sync.

The `ResyncPeriod` it's supposed to be at least tens of minutes, not a short time. So effectively, it should be infinity for the test.

Reply:
Ok, makes sense.

Still I am lost where is the issue, do we need to write the test in a different way or there is a bug which causes the Informer cache to not have the Pod under load?

Reply:
The function `addPod` is called by the informer cache itself after it has added the Pod. So if you see `Lowered expectations`, then it means that the Pod should be in the cache.

I'm not sure what we can be missing.

Reply:
It works somewhat better now, but it still fails once every ~300 test runs.
With the expectations, the `podStore` returns a pod but now occassionally it fails in this block https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L292-L295

`jm.resolveControllerRef(pod.Namespace, controllerRef)` returns nil and the function returns, so the creation is not observed and it fails in the poll loop in the test when waiting for the `manager.expectations.SatisfiedExpectations` to return true.

Reply:
maybe a "brute force" approach is ok.... you have access to the listers, just make sure they are not empty before proceeding. I guess we can do the same for making sure the pod is in the lister, instead of using the expectations.